### PR TITLE
[JSC][uDFG] Introduce unlinked CallLinkInfo in DFG

### PR DESCRIPTION
**Pull request by Constellation**
#### e8d10964fe9c69be268e5160f9839933f3d01965
<pre>
[JSC][uDFG] Introduce unlinked CallLinkInfo in DFG
<a href="https://bugs.webkit.org/show_bug.cgi?id=246056">https://bugs.webkit.org/show_bug.cgi?id=246056</a>
&lt;rdar://100788330&gt;

Reviewed by Justin Michaud.

This patch introduces unlinked CallLinkInfo into DFG. We do the similar thing to unlinked StructureStubInfo introduction
in DFG: Adding std::variant CompileTimeCallLinkInfo to handle unlinked and linked ones easily. We unify BaselineJIT&apos;s DataIC
CallLinkInfo code generation code with this new style so that we simplify unlinked CallLinkInfo usage.
The key part is that we no longer emit a pointers to CallLinkInfo in DFG if isUnlinked() (which is tested in JSC stress tests)
is true. In that case, we keep UnlinkedCallLinkInfo information in DFG::JITCode. And when running it, we materialize DFG::JITData,
which includes linked version of CallLinkInfo created from UnlinkedCallLinkInfo, so we share the unlinked DFG code while we have
per CodeBlock&apos;s linked data in DFG::JITData, which is pinned in a certain callee-save register in DFG.

* Source/JavaScriptCore/bytecode/AccessCase.cpp:
(JSC::AccessCase::generateImpl):
* Source/JavaScriptCore/bytecode/CallLinkInfo.cpp:
(JSC::CallLinkInfo::emitFastPath):
(JSC::CallLinkInfo::emitTailCallFastPath):
(JSC::CallLinkInfo::emitSlowPath):
(JSC::OptimizingCallLinkInfo::initializeFromDFGUnlinkedCallLinkInfo):
* Source/JavaScriptCore/bytecode/CallLinkInfo.h:
(JSC::UnlinkedCallLinkInfo::setCodeLocations):
(JSC::BaselineUnlinkedCallLinkInfo::setUpCall):
(JSC::BaselineUnlinkedCallLinkInfo::setFrameShuffleData):
* Source/JavaScriptCore/bytecode/ProxyObjectAccessCase.cpp:
(JSC::ProxyObjectAccessCase::emit):
* Source/JavaScriptCore/dfg/DFGCommonData.h:
(JSC::DFG::CommonData::addCallLinkInfo): Deleted.
* Source/JavaScriptCore/dfg/DFGJITCode.cpp:
(JSC::DFG::JITData::JITData):
* Source/JavaScriptCore/dfg/DFGJITCode.h:
(JSC::DFG::UnlinkedCallLinkInfo::setUpCall):
(JSC::DFG::UnlinkedCallLinkInfo::setFrameShuffleData):
(JSC::DFG::JITData::create):
* Source/JavaScriptCore/dfg/DFGJITCompiler.cpp:
(JSC::DFG::JITCompiler::link):
(JSC::DFG::JITCompiler::addCallLinkInfo):
* Source/JavaScriptCore/dfg/DFGJITCompiler.h:
(JSC::DFG::JITCompiler::addJSCall):
(JSC::DFG::JITCompiler::JSCallRecord::JSCallRecord):
* Source/JavaScriptCore/dfg/DFGPlan.cpp:
(JSC::DFG::Plan::finalizeJITData):
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT32_64.cpp:
(JSC::DFG::SpeculativeJIT::emitCall):
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT64.cpp:
(JSC::DFG::SpeculativeJIT::emitCall):
* Source/JavaScriptCore/ftl/FTLLowerDFGToB3.cpp:
(JSC::FTL::DFG::LowerDFGToB3::compileCompareStrictEq):
* Source/JavaScriptCore/ftl/FTLState.cpp:
(JSC::FTL::State::addCallLinkInfo):
* Source/JavaScriptCore/ftl/FTLState.h:
* Source/JavaScriptCore/jit/BaselineJITCode.h:
* Source/JavaScriptCore/jit/JIT.cpp:
(JSC::JIT::addUnlinkedCallLinkInfo):
(JSC::JIT::link):
* Source/JavaScriptCore/jit/JIT.h:
* Source/JavaScriptCore/jit/JITCall.cpp:
(JSC::JIT::compileCallDirectEval):
(JSC::JIT::compileTailCall):
(JSC::JIT::compileOpCall):
(JSC::JIT::compileOpCallSlowCase):
* Source/JavaScriptCore/wasm/js/WasmToJS.cpp:
(JSC::Wasm::wasmToJS):

Canonical link: <a href="https://commits.webkit.org/255183@main">https://commits.webkit.org/255183@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/f3368d160d7ce605c58eb3a1cbe157f7a3662d04

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/91588 "9 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/765 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/22199 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/101290 "Built successfully") | [❌ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/161354 "Reverted pull request changes (failure)") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/764 "Built successfully") | [  ~~🛠 mac-debug~~](https://ews-build.webkit.org/#/builders/71/builds/29434 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/83906 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/97632 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/97246 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/463 "Passed tests") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/61/builds/78259 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/27420 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/82387 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/82034 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/70457 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/83017 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/35711 "Built successfully") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/16064 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/78062 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/33469 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/17158 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/26959 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/3597 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/37306 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/61/builds/78259 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/80669 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/39212 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/36275 "Passed tests") | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/17675 "Passed tests") |

### CODE REVIEW

File: Source/JavaScriptCore/dfg/DFGJITCode.cpp [hunk 1]
```diff
@@ -51,6 +52,14 @@ JITData::JITData(const JITCode& jitCode, ExitVector&& exits)
             at(i) = &stubInfo;
             break;
         }
+        case LinkerIR::Type::CallLinkInfo: {
+            unsigned index = bitwise_cast<uintptr_t>(entry.pointer());
+            const UnlinkedCallLinkInfo& unlinkedCallLinkInfo = jitCode.m_unlinkedCallLinkInfos[index];
+            OptimizingCallLinkInfo& callLinkInfo = m_callLinkInfos[index];
+            callLinkInfo.initializeFromDFGUnlinkedCallLinkInfo(vm, unlinkedCallLinkInfo);
```
Comment (justinmichaud) [inline Source/JavaScriptCore/dfg/DFGJITCode.cpp:59]: Maybe I'm not understanding how this is supposed to be initialized, but could this be a placement new instead?

File: Source/JavaScriptCore/jit/JITCall.cpp [hunk 2]
```diff
@@ -270,21 +267,23 @@ void JIT::compileOpCall(const JSInstruction* instruction, unsigned callLinkInfoI
     emitGetVirtualRegister(callee, BaselineJITRegisters::Call::calleeJSR);
     storeValue(BaselineJITRegisters::Call::calleeJSR, Address(stackPointerRegister, CallFrameSlot::callee * static_cast<int>(sizeof(Register)) - sizeof(CallerFrameAndPC)));
 
-    if (compileCallDirectEval(bytecode))
+    if (opcodeID == op_call_direct_eval) {
```
Comment (justinmichaud) [inline Source/JavaScriptCore/jit/JITCall.cpp:270]: This is unrelated, right?

File: Source/JavaScriptCore/dfg/DFGSpeculativeJIT64.cpp [hunk 3]
```diff
@@ -1002,70 +1003,76 @@ void SpeculativeJIT::emitCall(Node* node)
...
     
     m_jit.emitStoreCallSiteIndex(callSite);
     
     CCallHelpers::JumpList slowCases;
     std::optional<CCallHelpers::Jump> done;
+    if (m_graph.m_plan.isUnlinked())
+        m_jit.loadLinkableConstant(callLinkInfoConstant, callLinkInfoGPR);
     if (isTail) {
-        slowCases = callLinkInfo->emitTailCallFastPath(m_jit, calleeGPR, callLinkInfoGPR, scopedLambda<void()>([&]{
+        slowCases = CallLinkInfo::emitTailCallFastPath(m_jit, callLinkInfo, calleeGPR, callLinkInfoGPR, scopedLambda<void()>([&, callLinkInfo = callLinkInfo]{
```
Comment (justinmichaud) [inline Source/JavaScriptCore/dfg/DFGSpeculativeJIT64.cpp:1062]: Will clang make sure we don't accidentally reference a stack-allocated object?

File: Source/JavaScriptCore/dfg/DFGJITCompiler.cpp [hunk 4]
```diff
@@ -800,6 +809,20 @@ std::tuple<CompileTimeStructureStubInfo, JITCompiler::LinkableConstant> JITCompi
...
 }
 
+std::tuple<CompileTimeCallLinkInfo, JITCompiler::LinkableConstant> JITCompiler::addCallLinkInfo(CodeOrigin codeOrigin)
+{
+    if (m_graph.m_plan.isUnlinked()) {
+        void* unlinkedCallLinkInfoIndex = bitwise_cast<void*>(static_cast<uintptr_t>(m_unlinkedCallLinkInfos.size()));
+        UnlinkedCallLinkInfo* callLinkInfo = &m_unlinkedCallLinkInfos.alloc();
+        callLinkInfo->codeOrigin = codeOrigin;
+        LinkerIR::Constant callLinkInfoIndex = addToConstantPool(LinkerIR::Type::CallLinkInfo, unlinkedCallLinkInfoIndex);
+        return std::tuple { callLinkInfo, LinkableConstant(callLinkInfoIndex) };
```
Comment (justinmichaud) [inline Source/JavaScriptCore/dfg/DFGJITCompiler.cpp:819]: So this adds a callLinkInfo to m_unlinkedCallLinkInfos, but why can't it add it to m_jitCode directly and avoid the move above?

Comment (Constellation) [reply Source/JavaScriptCore/dfg/DFGJITCompiler.cpp:819]: Because we would like to use FixedVector in JITCode to save memory as much as possible. JITCode is alive so long, but JITCompiler dies quickly. We allocate the backing vector in a faster way in JITCompiler and when finally building JITCode, copy them with the smallest size to keep them small.

File: Source/JavaScriptCore/dfg/DFGSpeculativeJIT64.cpp (hunk 3 above)
Comment (Constellation) [reply Source/JavaScriptCore/dfg/DFGSpeculativeJIT64.cpp:1062]: Not sure but it is explicitly said that this is stack only functor here (scopedLambda)

File: Source/JavaScriptCore/jit/JITCall.cpp (hunk 2 above)
Comment (Constellation) [reply Source/JavaScriptCore/jit/JITCall.cpp:270]: Cleaning up the flow to make it explicit that op_call_direct_eval does not use CallLinkInfo. (since we return in this branch)

File: Source/JavaScriptCore/dfg/DFGJITCode.cpp (hunk 1 above)
Comment (Constellation) [reply Source/JavaScriptCore/dfg/DFGJITCode.cpp:59]: Because it is already initialized, it is not great that we re-initialize the already initialized memory region (UB).

File: Source/JavaScriptCore/dfg/DFGSpeculativeJIT64.cpp (hunk 3 above)
Comment (justinmichaud) [reply Source/JavaScriptCore/dfg/DFGSpeculativeJIT64.cpp:1062]: Why do we capture callLinkInfo explicitly here?

**Review by justinmichaud — APPROVED** [review]
r=me

File: Source/JavaScriptCore/dfg/DFGSpeculativeJIT64.cpp (hunk 3 above)
Comment (Constellation) [reply Source/JavaScriptCore/dfg/DFGSpeculativeJIT64.cpp:1062]: This is because of a bit weird C++ semantics, where we need an explicit capture for destructuring-assignment-created variables.

**Comment by Constellation** [comment]
After 13 hours, debug build didn't happen.
